File: src/Sphere.java
import java.awt.Color;
import java.util.Optional;

public class Sphere implements Object3D {
    private Vector3D center;
    private float radius;
    private Color color;

    public Sphere(Vector3D center, float radius, Color color) {
        this.center = center;
        this.radius = radius;
        this.color = color;
    }

    public Vector3D getCenter() {
        return center;
    }

    public float getRadius() {
        return radius;
    }
    @Override
    public Color getColor() {
        return color;
    }

    @Override
    public Optional<Intersection> isHit(Ray ray) {
        Vector3D oc = ray.getPos().add(center.mul(-1));
        float a = ray.getDir().dot(ray.getDir());
        float b = 2.0f * oc.dot(ray.getDir());
        float c = oc.dot(oc) - radius * radius;
        float discriminant = b * b - 4 * a * c;
        if (discriminant < 0) {
            return Optional.empty();
        } else {
            float t = (-b - (float) Math.sqrt(discriminant)) / (2.0f * a);
            if (t > 0) {
                return Optional.of(new Intersection(this, t));
            } else {
                return Optional.empty();
            }
        }
    }

}

File: src/sphere.cpp
module sphere;
import std;
namespace cg
{

std::optional<hitevent> get_ray_sphere_hit(const sphere& s, ray r, interval i)
{
    const auto dir_to_sphere = s.pos - r.pos;
    auto a = glm::length2(r.dir);
    auto h = glm::dot(r.dir, dir_to_sphere);
    auto c = glm::length2(dir_to_sphere) - s.radius * s.radius;

    auto discriminant = h * h - a * c;

    if (discriminant < 0.)
    {
        return std::nullopt;
    }

    const double sqrt_discriminant = std::sqrt(discriminant);
    double t = (h - sqrt_discriminant) / a;
    if (t < 0. || !i.contains(t))
    {
        t = (h + sqrt_discriminant) / a;
        if (t < 0. || !i.contains(t))
            return std::nullopt;
    }
    auto normal = glm::normalize((r.at(t) - s.pos) / s.radius);
    if (glm::dot(normal, r.dir) > 0.)
    {
        normal = -normal;
    }
    return hitevent{.t = t,
                    .p = r.at(t),
                    .normal = normal,
                    .uv = vec2{0., 0.},
                    .m_id = s.material_id};
}

} // namespace cg
